File: scripts/seo.py
import requests
import time
import json
import os
from urllib.parse import urlparse

USER_ID = "5f9T1QuvWvbnkogeUqfk7lmUpsm1"
SAVED_ITEM_ID = "ujef8PYUuWgAtDcYR43aNF"
API_KEY = "1facf19033a348ffb0724746338ef0cb"

FOCUS_AREA = "Content Optimization (Hero, Services, Header, Meta, FAQ, CTAs, etc.)"


def get_url_directory_name(url: str):
    """Extract a clean directory name from URL"""
    parsed = urlparse(url)
    domain = parsed.netloc.replace('www.', '')
    # Replace dots and other special chars with underscores
    clean_domain = domain.replace('.', '_').replace('-', '_')
    return clean_domain


def create_url_directory(url: str):
    """Create directory structure: files/<domain>/"""
    domain_dir = get_url_directory_name(url)
    files_dir = os.path.join("files", domain_dir)
    os.makedirs(files_dir, exist_ok=True)
    return files_dir


def save_results_to_files(result: dict, url: str):
    """Save both JSON results and markdown content to organized file structure"""
    files_dir = create_url_directory(url)
    
    # Save JSON results
    json_path = os.path.join(files_dir, "result.json")
    try:
        with open(json_path, 'w', encoding='utf-8') as f:
            json.dump(result, f, indent=2, ensure_ascii=False)
        print(f"💾 JSON results saved to: {json_path}")
    except Exception as e:
        print(f"❌ Error saving JSON file: {e}")
    
    # Save markdown content if available
    outputs = result.get("outputs", {})
    markdown_content = outputs.get("output", "")
    
    if markdown_content and (markdown_content.strip().startswith('#') or '##' in markdown_content):
        md_path = os.path.join(files_dir, "content.md")
        try:
            with open(md_path, 'w', encoding='utf-8') as f:
                f.write(markdown_content)
            print(f"💾 Markdown content saved to: {md_path}")
        except Exception as e:
            print(f"❌ Error saving markdown file: {e}")
    else:
        print("⚠️  No markdown content found to save")


def start_pipeline(url: str):
    url = f"https://api.gumloop.com/api/v1/start_pipeline?user_id={USER_ID}&saved_item_id={SAVED_ITEM_ID}"
    payload = {"url":url, "focus_area":FOCUS_AREA}
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {API_KEY}"
    }
    response = requests.request("POST", url, json=payload, headers=headers)
    return response.json()


def get_run_result(run_id: str):
    url = f"https://api.gumloop.com/api/v1/get_pl_run?user_id={USER_ID}&run_id={run_id}"
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    response = requests.request("GET", url, headers=headers)
    return response.json()


def poll_job_status(run_id: str, poll_interval: int = 2, url: str = None):
    """
    Poll job status every poll_interval seconds until completion.
    Pretty prints the output when the job is done.
    If url is provided, saves results to organized file structure.
    """
    print(f"Polling job status for run_id: {run_id}")
    print(f"Checking every {poll_interval} seconds...")
    print("-" * 50)
    
    while True:
        try:
            result = get_run_result(run_id)
            state = result.get("state", "UNKNOWN")
            
            print(f"[{time.strftime('%H:%M:%S')}] Job state: {state}")
            
            if state == "DONE":
                print("\n" + "="*60)
                print("JOB COMPLETED SUCCESSFULLY!")
                print("="*60)
                
                # Save results to organized file structure if URL is provided
                if url:
                    print("\n💾 Saving results to organized file structure...")
                    save_results_to_files(result, url)
                
                outputs = result.get("outputs", {})
                if outputs:
                    print("\n📊 OUTPUTS:")
                    print("-" * 30)
                    for key, value in outputs.items():
                        print(f"\n🔑 {key}:")
                        if isinstance(value, str):
                            # Pretty print string output
                            print(value)
                        else:
                            # Pretty print JSON output
                            print(json.dumps(value, indent=2))
                else:
                    print("\n⚠️  No outputs found in the result")
                
                # Also show some job statistics
                print(f"\n📈 JOB STATISTICS:")
                print(f"   • Credit Cost: {result.get('credit_cost', 'N/A')}")
                print(f"   • Child Run Credit Cost: {result.get('child_run_credit_cost', 'N/A')}")
                print(f"   • Node Executions: {result.get('node_executions', 'N/A')}")
                print(f"   • Created: {result.get('created_ts', 'N/A')}")
                print(f"   • Finished: {result.get('finished_ts', 'N/A')}")
                
                break
                
            elif state == "FAILED":
                print("\n" + "="*60)
                print("❌ JOB FAILED!")
                print("="*60)
                print(f"Error details: {result}")
                break
                
            elif state == "RUNNING":
                # Show progress from logs if available
                logs = result.get("log", [])
                if logs:
                    recent_logs = logs[-3:]  # Show last 3 log entries
                    print("   Recent activity:")
                    for log_entry in recent_logs:
                        # Clean up ANSI escape codes for better readability
                        clean_log = log_entry.replace('\x1b[34m', '').replace('\x1b[32m', '').replace('\x1b[0m', '')
                        if '__system__:' in clean_log:
                            continue  # Skip system logs
                        print(f"     • {clean_log}")
                
            else:
                print(f"   Unknown state: {state}")
            
            time.sleep(poll_interval)
            
        except KeyboardInterrupt:
            print("\n\n⏹️  Polling interrupted by user")
            break
        except Exception as e:
            print(f"\n❌ Error polling job status: {e}")
            time.sleep(poll_interval)


def save_markdown_output(run_id: str, output_content: str = None, url: str = None):
    """
    Save markdown output from a completed job to a file.
    If output_content is not provided, fetches the latest result.
    If url is provided, saves to organized file structure.
    """
    if output_content is None:
        result = get_run_result(run_id)
        outputs = result.get("outputs", {})
        output_content = outputs.get("output", "")
    
    if not output_content or not (output_content.strip().startswith('#') or '##' in output_content):
        print("❌ No markdown content found in outputs")
        return None
    
    if url:
        # Use organized file structure
        files_dir = create_url_directory(url)
        filename = os.path.join(files_dir, "content.md")
    else:
        # Use simple filename
        filename = f"{run_id}.md"
    
    try:
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(output_content)
        print(f"💾 Markdown report saved to: {filename}")
        return filename
    except Exception as e:
        print(f"❌ Error saving markdown file: {e}")
        return None


import argparse
import sys

def main():
    parser = argparse.ArgumentParser(description='SEO Pipeline CLI')
    subparsers = parser.add_subparsers(dest="command", required=True)

    # 'run' command to start pipeline with url
    run_parser = subparsers.add_parser('run', help='Start pipeline with a URL')
    run_parser.add_argument('url', type=str, help='URL to process')
    run_parser.add_argument('--poll', action='store_true', help='Automatically poll for results after starting')

    # 'results' command to get results with run id
    results_parser = subparsers.add_parser('results', help='Get results for a given run ID')
    results_parser.add_argument('run_id', type=str, help='Run ID to fetch results for')

    # 'poll' command to poll job status
    poll_parser = subparsers.add_parser('poll', help='Poll job status until completion')
    poll_parser.add_argument('run_id', type=str, help='Run ID to poll')
    poll_parser.add_argument('--interval', type=int, default=2, help='Polling interval in seconds (default: 2)')
    poll_parser.add_argument('--url', type=str, help='URL to organize files by domain')

    # 'save' command to save markdown output
    save_parser = subparsers.add_parser('save', help='Save markdown output from completed job')
    save_parser.add_argument('run_id', type=str, help='Run ID to save output for')
    save_parser.add_argument('--url', type=str, help='URL to organize files by domain')

    args = parser.parse_args()

    if args.command == 'run':
        result = start_pipeline(args.url)
        run_id = result.get("run_id")
        if run_id:
            print(f"Pipeline started with run_id: {run_id}")
            if args.poll:
                print("\nStarting automatic polling...")
                poll_job_status(run_id, url=args.url)
            else:
                print(f"Use 'python seo.py poll {run_id} --url {args.url}' to monitor progress")
        else:
            print("Failed to retrieve run_id. Response:", result)
            sys.exit(1)
    elif args.command == 'results':
        result = get_run_result(args.run_id)
        print(result)
    elif args.command == 'poll':
        poll_job_status(args.run_id, args.interval, args.url)
    elif args.command == 'save':
        save_markdown_output(args.run_id, url=args.url)
    else:
        parser.print_help()

if __name__ == '__main__':
    main()



# url = f"https://api.gumloop.com/api/v1/start_pipeline?user_id={USER_ID}&saved_item_id={SAVED_ITEM_ID}"
# payload = {"URL":"https://www.vipcarcare.com.au"}
# headers = {
# "Content-Type": "application/json",
# "Authorization": f"Bearer {API_KEY}"
# }
# response = requests.request("POST", url, json=payload, headers=headers)
# result = response.json()


# url = "https://api.gumloop.com/api/v1/get_pl_run"
# params = {
#     "run_id": result["run_id"],
#     "user_id": USER_ID
# }
# headers = {
#     "Authorization": f"Bearer {API_KEY}"
# }

# response = requests.get(url, params=params, headers=headers)
# result = response.json()

# print(result)

File: convex/seo.ts
import { action } from "./_generated/server";
import { v } from "convex/values";

// Gumloop API response types
interface StartPipelineResponse {
  run_id: string;
  state?: string;
}

interface RunResultResponse {
  state: "PENDING" | "RUNNING" | "DONE" | "FAILED";
  outputs?: {
    output?: string; // The markdown content
  };
  credit_cost?: number;
  child_run_credit_cost?: number;
  node_executions?: number;
  created_ts?: string;
  finished_ts?: string;
  error?: string;
}

export interface SEOAnalysisResult {
  success: boolean;
  url: string;
  markdown?: string;
  runId?: string;
  error?: string;
}

/**
 * Start a Gumloop SEO analysis pipeline
 */
async function startPipeline(
  url: string,
  apiKey: string,
  userId: string,
  savedItemId: string,
  focusArea: string
): Promise<StartPipelineResponse> {
  const apiUrl = `https://api.gumloop.com/api/v1/start_pipeline?user_id=${userId}&saved_item_id=${savedItemId}`;

  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${apiKey}`,
  };

  const payload = {
    url: url,
    focus_area: focusArea,
  };

  try {
    console.log(`[SEO] Starting pipeline for URL: ${url}`);
    console.log(`[SEO] Focus area: ${focusArea}`);

    const response = await fetch(apiUrl, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`[SEO] Start pipeline failed: ${response.status} - ${errorText}`);
      throw new Error(`Failed to start pipeline: ${response.status}`);
    }

    const data = (await response.json()) as StartPipelineResponse;
    console.log(`[SEO] Pipeline started with run_id: ${data.run_id}`);

    return data;
  } catch (error) {
    console.error("[SEO] Error starting pipeline:", error);
    throw error;
  }
}

/**
 * Get the current result/status of a pipeline run
 */
async function getRunResult(
  runId: string,
  apiKey: string,
  userId: string
): Promise<RunResultResponse> {
  const apiUrl = `https://api.gumloop.com/api/v1/get_pl_run?user_id=${userId}&run_id=${runId}`;

  const headers = {
    Authorization: `Bearer ${apiKey}`,
  };

  try {
    const response = await fetch(apiUrl, {
      method: "GET",
      headers,
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`[SEO] Get run result failed: ${response.status} - ${errorText}`);
      throw new Error(`Failed to get run result: ${response.status}`);
    }

    const data = (await response.json()) as RunResultResponse;
    return data;
  } catch (error) {
    console.error("[SEO] Error getting run result:", error);
    throw error;
  }
}

/**
 * Poll the pipeline status until completion or timeout
 * Max duration: 180 seconds (3 minutes)
 * Poll interval: 5 seconds
 */
async function pollUntilComplete(
  runId: string,
  apiKey: string,
  userId: string,
  maxDurationSeconds: number = 180,
  pollIntervalSeconds: number = 5
): Promise<RunResultResponse> {
  const startTime = Date.now();
  const maxDurationMs = maxDurationSeconds * 1000;
  const pollIntervalMs = pollIntervalSeconds * 1000;
  const maxAttempts = Math.ceil(maxDurationSeconds / pollIntervalSeconds);

  let attempt = 0;

  console.log(`[SEO] Starting polling (max ${maxDurationSeconds}s, interval ${pollIntervalSeconds}s)`);

  while (true) {
    attempt++;
    const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);

    try {
      const result = await getRunResult(runId, apiKey, userId);
      const state = result.state;

      console.log(`[SEO] [${attempt}/${maxAttempts}] State: ${state} (${elapsed}s elapsed)`);

      if (state === "DONE") {
        console.log(`[SEO] ✓ Pipeline completed successfully (${elapsed}s total)`);
        if (result.outputs?.output) {
          console.log(`[SEO] Markdown content: ${result.outputs.output.length} characters`);
        }
        return result;
      }

      if (state === "FAILED") {
        console.error(`[SEO] ✗ Pipeline failed:`, result.error || "Unknown error");
        return result;
      }

      // Check timeout
      if (Date.now() - startTime >= maxDurationMs) {
        console.warn(`[SEO] ⚠️ Timeout after ${elapsed}s, pipeline still ${state}`);
        throw new Error(`Pipeline timeout after ${maxDurationSeconds}s`);
      }

      // Wait before next poll
      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    } catch (error) {
      const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
      console.error(`[SEO] Error during polling (${elapsed}s):`, error);
      throw error;
    }
  }
}

/**
 * Main action to fetch SEO analysis for a website
 * This is called from the AI spec generation flow in parallel with reviews and website scraping
 */
export const fetchSEOAnalysis = action({
  args: {
    websiteUrl: v.string(),
  },
  handler: async (_, args): Promise<SEOAnalysisResult> => {
    console.log(`\n========================================`);
    console.log(`[fetchSEOAnalysis] Starting for: "${args.websiteUrl}"`);
    console.log(`========================================\n`);

    const apiKey = process.env.GUMLOOP_API_KEY;
    const userId = process.env.GUMLOOP_USER_ID;
    const savedItemId = process.env.GUMLOOP_SAVED_ITEM_ID;
    const focusArea = process.env.GUMLOOP_FOCUS_AREA || "Content Optimization (Hero, Services, Header, Meta, FAQ, CTAs, etc.)";

    if (!apiKey || !userId || !savedItemId) {
      console.error("[fetchSEOAnalysis] ERROR: Missing Gumloop configuration");
      console.error(`- GUMLOOP_API_KEY: ${apiKey ? "set" : "missing"}`);
      console.error(`- GUMLOOP_USER_ID: ${userId ? "set" : "missing"}`);
      console.error(`- GUMLOOP_SAVED_ITEM_ID: ${savedItemId ? "set" : "missing"}`);
      return {
        success: false,
        url: args.websiteUrl,
        error: "Gumloop API configuration not found",
      };
    }

    try {
      // Step 1: Start the pipeline
      const startResult = await startPipeline(
        args.websiteUrl,
        apiKey,
        userId,
        savedItemId,
        focusArea
      );

      if (!startResult.run_id) {
        console.error("[fetchSEOAnalysis] FAILED: No run_id returned");
        return {
          success: false,
          url: args.websiteUrl,
          error: "Failed to start SEO pipeline",
        };
      }

      // Step 2: Poll until completion (max 3 minutes)
      const result = await pollUntilComplete(startResult.run_id, apiKey, userId, 180, 5);

      if (result.state === "FAILED") {
        console.log(`[fetchSEOAnalysis] FAILED: Pipeline failed`);
        return {
          success: false,
          url: args.websiteUrl,
          runId: startResult.run_id,
          error: result.error || "SEO analysis pipeline failed",
        };
      }

      if (result.state !== "DONE") {
        console.log(`[fetchSEOAnalysis] INCOMPLETE: Pipeline state: ${result.state}`);
        return {
          success: false,
          url: args.websiteUrl,
          runId: startResult.run_id,
          error: "SEO analysis did not complete in time",
        };
      }

      // Step 3: Extract markdown content
      const markdown = result.outputs?.output;

      if (!markdown || markdown.trim().length === 0) {
        console.log(`[fetchSEOAnalysis] WARNING: No markdown content in output`);
        return {
          success: false,
          url: args.websiteUrl,
          runId: startResult.run_id,
          error: "No SEO analysis content generated",
        };
      }

      console.log(`\n[fetchSEOAnalysis] SUCCESS Summary:`);
      console.log(`- Website: ${args.websiteUrl}`);
      console.log(`- Run ID: ${startResult.run_id}`);
      console.log(`- Markdown length: ${markdown.length} characters`);
      console.log(`- Credit cost: ${result.credit_cost || "N/A"}`);
      console.log(`========================================\n`);

      return {
        success: true,
        url: args.websiteUrl,
        runId: startResult.run_id,
        markdown,
      };
    } catch (error) {
      console.error("[fetchSEOAnalysis] ERROR:", error);
      return {
        success: false,
        url: args.websiteUrl,
        error: error instanceof Error ? error.message : "Unknown error",
      };
    }
  },
});
